src/app/ReadClient.cpp

```diff
@@ -77,95 +78,112 @@ void ReadClient::MoveToState(const ClientState aTargetState)
-    ChipLogDetail(DataManagement, "%s: Client[%u] [%5.5s]", __func__,
-                  InteractionModelEngine::GetInstance()->GetReadClientArrayIndex(this), GetStateStr());
-    VerifyOrExit(ClientState::Initialized == mState, err = CHIP_ERROR_INCORRECT_STATE);
-    VerifyOrExit(mpDelegate != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
-
-    // Discard any existing exchange context. Effectively we can only have one exchange per ReadClient
-    // at any one time.
-    AbortExistingExchangeContext();
+    ReadRequest::Builder request;
+    err = request.Init(&aWriter);
+    SuccessOrExit(err);
 
+    if (aEventPathParamsListSize != 0 && apEventPathParamsList != nullptr)
     {
-        System::PacketBufferTLVWriter writer;
-        ReadRequest::Builder request;
-
-        msgBuf = System::PacketBufferHandle::New(kMaxSecureSduLengthBytes);
-        VerifyOrExit(!msgBuf.IsNull(), err = CHIP_ERROR_NO_MEMORY);
-
-        writer.Init(std::move(msgBuf));
-
-        err = request.Init(&writer);
-        SuccessOrExit(err);
-
-        if (aEventPathParamsListSize != 0 && apEventPathParamsList != nullptr)
+        EventPathList::Builder & eventPathListBuilder = request.CreateEventPathListBuilder();
+        EventPath::Builder eventPathBuilder           = eventPathListBuilder.CreateEventPathBuilder();
+        for (size_t eventIndex = 0; eventIndex < aEventPathParamsListSize; ++eventIndex)
         {
-            EventPathList::Builder & eventPathListBuilder = request.CreateEventPathListBuilder();
-            EventPath::Builder eventPathBuilder           = eventPathListBuilder.CreateEventPathBuilder();
-            for (size_t eventIndex = 0; eventIndex < aEventPathParamsListSize; ++eventIndex)
-            {
-                EventPathParams eventPath = apEventPathParamsList[eventIndex];
-                eventPathBuilder.NodeId(eventPath.mNodeId)
-                    .EventId(eventPath.mEventId)
-                    .EndpointId(eventPath.mEndpointId)
-                    .ClusterId(eventPath.mClusterId)
-                    .EndOfEventPath();
-                SuccessOrExit(err = eventPathBuilder.GetError());
-            }
+            EventPathParams eventPath = apEventPathParamsList[eventIndex];
+            eventPathBuilder.NodeId(eventPath.mNodeId)
+                .EventId(eventPath.mEventId)
+                .EndpointId(eventPath.mEndpointId)
+                .ClusterId(eventPath.mClusterId)
+                .EndOfEventPath();
+            SuccessOrExit(err = eventPathBuilder.GetError());
+        }
 
-            eventPathListBuilder.EndOfEventPathList();
-            SuccessOrExit(err = eventPathListBuilder.GetError());
+        eventPathListBuilder.EndOfEventPathList();
+        SuccessOrExit(err = eventPathListBuilder.GetError());
 
-            if (aEventNumber != 0)
-            {
-                // EventNumber is optional
-                request.EventNumber(aEventNumber);
-            }
+        if (aEventNumber != 0)
```

Comment:
should we have a constant for the '0' or is 0 good enough as unique? something like kInvalidEventNumber?